grader-notebook/grading/utils.py

```diff
@@ -0,0 +1,146 @@
+    of the nbgrader application and use its config.
+    """
+    grader_app = NbGrader()
+    grader_app.initialize(argv=[])
+    grader_app.notebook_dir = "."  # TODO: needed?
+
+    return grader_app.config
+
+
+def _run_app(app):
+    handler = logging.StreamHandler(sys.stdout)
+    formatter = logging.Formatter("[%(levelname)s] %(message)s")
+    handler.setFormatter(formatter)
+    app.log.addHandler(handler)
+
+    app.start()
+
+    app.log.removeHandler(handler)
+
+
+def autograde(assignment_slug, student_slug):
+    c = get_config()
+    c.CourseDirectory.assignment_id = assignment_slug
+    c.CourseDirectory.student_id = student_slug
+
+    coursedir = CourseDirectory(config=c)
+
+    with Gradebook(coursedir.db_url, coursedir.course_id) as gb:
+        # Store the submission in the database
+
+        # TODO: add name etc.?
+        gb.update_or_create_student(student_slug)
+        gb.update_or_create_submission(assignment_slug, student_slug)
+
+        # TODO: can we find out if grading already happened?
+        # can we then skip this part?
+
+    # Use the input in <course root>/submitted for the given assignment and studen
+    # and create a submission in the database as well as store the results in the database.
+    # Will need to evaluate the notebook! Needs database access.
+    app = Autograde(coursedir=coursedir, config=c)
+    app.force = True
+    app.create_student = True
+    _run_app(app)
+
+    # After that, the autograding is performed, but feedback is not present already
+
+
+def generate_feedback(assignment_slug, student_slug):
+    c = get_config()
+    c.CourseDirectory.assignment_id = assignment_slug
+    c.CourseDirectory.student_id = student_slug
+
+    coursedir = CourseDirectory(config=c)
+
+    # If manual grading is needed, we can not export it already!
+    with Gradebook(coursedir.db_url, coursedir.course_id) as gb:
+        submission = gb.find_submission(assignment_slug, student_slug)
+        if submission.needs_manual_grade:
+            print(
+                f"Not creating feedback for {assignment_slug} - {student_slug} as it needs manual grading."
+            )
+            return
+
+    # Generate the feedback and store it in the folder
+    #
+    # Will use the grades stored in the database to create
+    # a feedback.
+    app = GenerateFeedback(coursedir=coursedir, config=c)
+    app.force = True
+    _run_app(app)
+
+    # Works! Feedback present in content/feedback/...,
+
+    with Gradebook(coursedir.db_url, coursedir.course_id) as gb:
+        submission = gb.find_submission(assignment_slug, student_slug)
+
+        submission_dict = submission.to_dict()
+        submission_dict["notebooks"] = [n.to_dict() for n in submission.notebooks]
+
+        return submission_dict
+
+
+def get_ungraded_submissions():
+    api_token = os.environ.get("GRADER_API_TOKEN")
+    rv = requests.get(
+        "http://jupyterhub:8000/services/learningpage/api/v1/ungraded",
```

Comment:
Not sure if it's a good idea to hardcode hostname and port here, but for now it's okay I guess.